src/graphql-types/ReasonApolloQuery.re

```diff
@@ -1,84 +1,92 @@
-module type InternalConfig = {let apolloClient: ApolloClient.generatedApolloClient;};
-
-module QueryFactory = (InternalConfig:InternalConfig) => {
-    external castResponse : string => {. "data": Js.Json.t } = "%identity";            
-    external asJsObject : 'a => Js.t({..}) = "%identity";
+open ReasonApolloTypes;
 
+module Get = (Config:ReasonApolloTypes.Config) => {    
     [@bs.module] external gql : ReasonApolloTypes.gql = "graphql-tag";
-    [@bs.module] external shallowEqual : (Js.t({..}), Js.t({..})) => bool = "fbjs/lib/shallowEqual";
-
+    [@bs.module "react-apollo"] external queryComponent : ReasonReact.reactClass = "Query";
     
     type response =
       | Loading
-      | Loaded(Js.Json.t)
-      | Failed(string);
+      | Error(apolloError)
+      | Data(Config.t)
+      | NoData;
+
+    type renderPropObj = {
+      data: response,
+      refetch: option(Js.Json.t) => Js.Promise.t(response),
+      fetchMore: (~variables: Js.Json.t) => Js.Promise.t(unit),
+      networkStatus: int,
+    };
 
-    type state = { 
-      response: response, 
-      variables: Js.Json.t
+    type renderPropObjJS = {
+      .
+      "loading": Js.boolean,
+      "data": Js.Nullable.t(Js.Json.t),
+      "error": Js.Nullable.t(apolloError),
+      "refetch":
+        [@bs.meth] (
+          Js.Null_undefined.t(Js.Json.t) => Js.Promise.t(renderPropObjJS)
+        ),
+      "networkStatus": int,
+      "variables": Js.Null_undefined.t(Js.Json.t),
+      "fetchMore": [@bs.meth] (apolloOptions => Js.Promise.t(unit)),
     };
 
-    type action =
-      | Result(string)
-      | Error(string);
+    let graphqlQueryAST = [@bs] gql(Config.query);
 
-    let sendQuery = (~query, ~reduce) => {
-      let _ =
-      Js.Promise.(
-        resolve(InternalConfig.apolloClient##query({
-          "query": [@bs] gql(query##query),
-          "variables": query##variables
-        }))
-        |> then_(
-             (value) => {
-               reduce(() => Result(value), ());
-               resolve()
-             }
-           )
-        |> catch(
-             (_value) => {
-               reduce(() => Error("an error happened"), ());
-               resolve()
-             }
-           )
-      );
+    let apolloDataToReason: renderPropObjJS => response =
+    apolloData =>
+      switch (
+        apolloData##loading |> Js.to_bool,
+        apolloData##data |> Js.Nullable.to_opt,
+        apolloData##error |> Js.Nullable.to_opt,
+      ) {
+      | (true, _, _) => Loading
+      | (false, Some(response), _) => Data(Config.parse(response))
+      | (false, _, Some(error)) => Error(error)
+      | (false, None, None) => NoData
+      };
+    
+    let convertJsInputToReason = (apolloData: renderPropObjJS) => {
```

Comment:
In the issue discussion, we'd talked about adding a `result` key with the computed variant and keeping `data`, `loading`, and `error` as straight pass-throughs from Apollo so that API consumers could construct their own variant logic if the default one provided didn't suit their needs.